File: project2/golang/cmd/main.go
package main

import (
	"context"
	"fmt"
	"net/http"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/sirupsen/logrus"
	"project2-golang/config"
	"project2-golang/provider/amqpx"
	"project2-golang/provider/messaging"
	"project2-golang/provider/metrics"
)

func main() {
	// Load configuration
	cfg, err := config.LoadConfig(".", "./config")
	if err != nil {
		logrus.WithError(err).Fatal("Failed to load configuration")
	}

	// Setup logger
	logger := logrus.New()
	logger.SetFormatter(&logrus.JSONFormatter{})
	
	level, err := logrus.ParseLevel(cfg.Log.Level)
	if err != nil {
		level = logrus.InfoLevel
	}
	logger.SetLevel(level)

	logger.WithFields(logrus.Fields{
		"app_name": cfg.App.Name,
		"version":  "1.0.0",
	}).Info("Starting application")

	// Initialize metrics
	appMetrics := metrics.NewMetrics()
	
	// Connect to RabbitMQ
	conn, err := amqpx.Dial(cfg.AMQP.GetDSN())
	if err != nil {
		logger.WithError(err).Fatal("Failed to connect to RabbitMQ")
	}
	defer conn.Close()
	
	// Update connection metrics
	appMetrics.SetAMQPConnections(1)

	// Create channel for publisher
	publisherChannel, err := conn.Channel()
	if err != nil {
		logger.WithError(err).Fatal("Failed to create publisher channel")
	}
	defer publisherChannel.Close()

	// Create channel for consumer
	consumerChannel, err := conn.Channel()
	if err != nil {
		logger.WithError(err).Fatal("Failed to create consumer channel")
	}
	defer consumerChannel.Close()

	// Setup publisher
	publisher := messaging.NewAMQPPublisher(publisherChannel)

	// Setup consumer
	consumer := messaging.NewAMQPConsumer(
		consumerChannel,
		publisher,
		appMetrics,
		cfg.AMQP.Concurrent,
		cfg.AMQP.PrefetchCount,
	)

	// Setup queue processor
	processor := messaging.NewQueueProcessor(
		consumer,
		cfg.Queues.InputQueue,
		cfg.Queues.OutputQueue,
	)

	// Create context
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// Start metrics collection
	go func() {
		ticker := time.NewTicker(5 * time.Second)
		defer ticker.Stop()
		
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				appMetrics.UpdateSystemMetrics()
			}
		}
	}()

	// Start simple metrics server
	go func() {
		http.Handle("/metrics", promhttp.Handler())
		logger.WithField("port", cfg.App.Port).Info("Starting metrics server")
		if err := http.ListenAndServe(fmt.Sprintf(":%d", cfg.App.Port), nil); err != nil {
			logger.WithError(err).Error("Metrics server error")
		}
	}()

	// Start consuming messages
	err = consumer.StartConsuming(ctx, cfg.Queues.InputQueue, processor)
	if err != nil {
		logger.WithError(err).Fatal("Failed to start consumer")
	}

	logger.WithFields(logrus.Fields{
		"input_queue":  cfg.Queues.InputQueue,
		"output_queue": cfg.Queues.OutputQueue,
		"concurrency":  cfg.AMQP.Concurrent,
	}).Info("Queue processor started successfully")

	// Wait for interrupt signal
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)

	<-sigChan
	logger.Info("Shutdown signal received")

	// Graceful shutdown
	shutdownCtx, shutdownCancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer shutdownCancel()

	if err := consumer.Stop(shutdownCtx); err != nil {
		logger.WithError(err).Error("Error during consumer shutdown")
	}

	cancel() // Cancel main context
	logger.Info("Application shutdown complete")
}

File: project2/rust/src/main.rs
mod amqp;
mod config;
mod messaging;
mod metrics;

use anyhow::Result;
use axum::{routing::get, Router};
use std::sync::Arc;
use tokio::{net::TcpListener, signal};
use tracing::{error, info, Level};
use tracing_subscriber::{fmt, prelude::*, EnvFilter};

use amqp::AMQPConnection;
use config::AppConfig;
use messaging::{AMQPConsumer, AMQPPublisher, QueueProcessor};
use metrics::Metrics;

#[tokio::main]
async fn main() -> Result<()> {
    // Initialize tracing
    let subscriber = tracing_subscriber::registry()
        .with(fmt::layer().json())
        .with(
            EnvFilter::builder()
                .with_default_directive(Level::INFO.into())
                .from_env_lossy(),
        );

    tracing::subscriber::set_global_default(subscriber)?;

    // Load configuration
    let config = AppConfig::load().map_err(|e| {
        error!("Failed to load configuration: {}", e);
        e
    })?;

    info!(
        app_name = %config.app.name,
        version = "1.0.0",
        "Starting application"
    );

    // Initialize metrics
    let app_metrics = Arc::new(Metrics::new());

    // Connect to RabbitMQ
    let connection = AMQPConnection::new(&config.amqp.url).await?;
    
    // Update connection metrics
    app_metrics.set_amqp_connections(1.0);

    // Create channels
    let publisher_channel = connection.create_channel().await?;
    let consumer_channel = connection
        .create_channel_with_qos(config.amqp.prefetch_count)
        .await?;

    // Setup publisher
    let publisher = AMQPPublisher::new(publisher_channel);

    // Setup consumer
    let consumer = AMQPConsumer::new(
        consumer_channel,
        publisher.clone(),
        app_metrics.clone(),
        config.amqp.concurrent,
    );

    // Setup queue processor
    let processor = QueueProcessor::new(
        consumer,
        config.queues.input_queue.clone(),
        config.queues.output_queue.clone(),
    );

    info!(
        input_queue = %config.queues.input_queue,
        output_queue = %config.queues.output_queue,
        concurrency = config.amqp.concurrent,
        http_port = config.app.port,
        "Queue processor started successfully"
    );

    // Start simple metrics server
    let metrics_clone = app_metrics.clone();
    let metrics_port = config.app.port;
    let metrics_handle = tokio::spawn(async move {
        let app = Router::new()
            .route("/metrics", get({
                let metrics = metrics_clone;
                || async move {
                    match metrics.encode_metrics() {
                        Ok(encoded) => encoded,
                        Err(e) => {
                            error!("Failed to encode metrics: {}", e);
                            "".to_string()
                        }
                    }
                }
            }));

        let addr = format!("0.0.0.0:{}", metrics_port);
        let listener = match TcpListener::bind(&addr).await {
            Ok(listener) => listener,
            Err(e) => {
                error!("Failed to bind metrics server: {}", e);
                return;
            }
        };
        
        info!("Metrics server listening on {}", addr);
        
        if let Err(e) = axum::serve(listener, app).await {
            error!("Metrics server error: {}", e);
        }
    });

    // Start consuming messages
    let consumer_handle = {
        let processor_clone = processor.clone();
        let consumer_clone = processor.consumer.clone();
        let input_queue = config.queues.input_queue.clone();
        tokio::spawn(async move {
            if let Err(e) = consumer_clone.start_consuming(&input_queue, processor_clone).await {
                error!("Consumer error: {}", e);
            }
        })
    };

    // Wait for shutdown signal
    info!("Application running. Press Ctrl+C to shutdown.");
    signal::ctrl_c().await?;
    info!("Shutdown signal received");

    // Graceful shutdown
    consumer_handle.abort();
    metrics_handle.abort();
    info!("Application shutdown complete");

    Ok(())
}